fraud/service_test.go

```diff
@@ -56,18 +60,334 @@ func TestService_Broadcast(t *testing.T) {
+	bServ := mdutils.Bserv()
+
+	// create first fraud service that will broadcast incorrect Fraud Proof
+	serviceA, store1 := createServiceWithHost(ctx, t, net.Hosts()[0])
+
+	h, err := store1.GetByHeight(ctx, 1)
+	require.NoError(t, err)
+
+	// create and break byzantine error
+	_, err = generateByzantineError(ctx, t, h, bServ)
+	require.Error(t, err)
+	var errByz *ipld.ErrByzantine
+	require.True(t, errors.As(err, &errByz))
+	errByz.Index = 2
+
+	fserviceA := serviceA.(*service)
+	require.NotNil(t, fserviceA)
+
+	// create second service that will receive and validate Fraud Proof
+	serviceB, _ := createServiceWithHost(ctx, t, net.Hosts()[1])
+
+	fserviceB := serviceB.(*service)
+	require.NotNil(t, fserviceB)
+
+	bl, err := pubsub.NewTimeCachedBlacklist(time.Hour * 1)
+	require.NoError(t, err)
+	// create pub sub in order to listen for Fraud Proof
+	psC, err := pubsub.NewGossipSub(ctx, net.Hosts()[2], // -> C
+		pubsub.WithMessageSignaturePolicy(pubsub.StrictNoSign), pubsub.WithBlacklist(bl))
+	require.NoError(t, err)
+
+	addrB := host.InfoFromHost(net.Hosts()[1]) // -> B
+
+	serviceC := NewService(psC, store1.GetByHeight, sync.MutexWrap(datastore.NewMapDatastore()))
+
+	sub0, err := net.Hosts()[0].EventBus().Subscribe(&event.EvtPeerIdentificationCompleted{})
+	require.NoError(t, err)
+	sub2, err := net.Hosts()[2].EventBus().Subscribe(&event.EvtPeerIdentificationCompleted{})
+	require.NoError(t, err)
+
+	// connect peers: A -> B -> C, so A and C are not connected to each other
+	require.NoError(t, net.Hosts()[0].Connect(ctx, *addrB)) // host[0] is A
+	require.NoError(t, net.Hosts()[2].Connect(ctx, *addrB)) // host[2] is C
+
+	// wait on both peer identification events
+	for i := 0; i < 2; i++ {
+		select {
+		case <-sub0.Out():
+		case <-sub2.Out():
+		case <-ctx.Done():
+			assert.FailNow(t, "timeout waiting for peers to connect")
+		}
+	}
+
+	// subscribe to BEFP
+	subsA, err := serviceA.Subscribe(BadEncoding)
+	require.NoError(t, err)
+	defer subsA.Cancel()
+
+	subsB, err := serviceB.Subscribe(BadEncoding)
+	require.NoError(t, err)
+	defer subsB.Cancel()
+
+	subsC, err := serviceC.Subscribe(BadEncoding)
+	require.NoError(t, err)
+	defer subsC.Cancel()
+
+	befp := CreateBadEncodingProof([]byte("hash"), uint64(h.Height), errByz)
+	// deregister validator in order to send Fraud Proof
+	fserviceA.pubsub.UnregisterTopicValidator(getSubTopic(BadEncoding)) //nolint:errcheck
+	// create a new validator for serviceB
+	fserviceB.pubsub.UnregisterTopicValidator(getSubTopic(BadEncoding)) //nolint:errcheck
+	f := func(ctx context.Context, from peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
+		msg.ValidatorData = befp
+		return pubsub.ValidationAccept
+	}
+	fserviceB.pubsub.RegisterTopicValidator(getSubTopic(BadEncoding), f) //nolint:errcheck
+	bin, err := befp.MarshalBinary()
+	require.NoError(t, err)
+	topic, ok := fserviceA.topics[BadEncoding]
+	require.True(t, ok)
+	// we cannot avoid sleep because it helps to avoid flakiness
```

Comment:
What is this sleep for though? What are you waiting on?

Reply:
libp2p specific sleeps 😞 
Test will fail without it